### PR TITLE
[AM][FEAT-35]: integrate min.io for object storage

### PR DESCRIPTION
**Pull request by tiescl**

### CODE REVIEW

**Review by bek-shoyatbek — APPROVED** [review]

File: src/modules/minio/minio.service.ts [hunk 1]
```diff
@@ -0,0 +1,81 @@
+import { Injectable, InternalServerErrorException } from '@nestjs/common';
+import { ConfigService } from '@nestjs/config';
+import { isProd } from 'src/shared/helpers';
+import * as Minio from 'minio';
+
+@Injectable()
+export class MinioService {
+  private minioClient: Minio.Client;
+  private bucketName: string;
```
Comment (farrooh) [inline src/modules/minio/minio.service.ts:9]: Create a type for bucket names

**Review by farrooh — CHANGES_REQUESTED** [review]

File: src/modules/minio/minio.service.ts (hunk 1 above)
Comment (tiescl) [reply src/modules/minio/minio.service.ts]: fixed

**Comment by Ismoil-JS** [comment]
@tiescl after resolving change status to resolved, and ask from teachers to review again

**Review by mukhammadjon-ss — APPROVED** [review]

File: src/modules/minio/minio.service.ts [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        },
+      ],
+    };
+
+    let prevPolicy: typeof bucketPolicyRead;
+    try {
+      const policy = await this.minioClient.getBucketPolicy(bucketName);
+      prevPolicy = JSON.parse(policy);
+    } catch (err) {
+      console.error(err.message);
```
Comment (farrooh) [inline src/modules/minio/minio.service.ts:50]: Please use Nest.js Logger instead. (For all logs)

File: src/modules/minio/minio.service.ts [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+        {
+          Effect: 'Allow',
+          Principal: '*',
+          Action: ['s3:GetObject'],
+          Resource: [`arn:aws:s3:::${bucketName}/*`],
+        },
+      ],
+    };
+
+    let prevPolicy: typeof bucketPolicyRead;
```
Comment (farrooh) [inline src/modules/minio/minio.service.ts:45]: It's better to give an initial value to ensure type safety. `... = bucketPolicyRead`;

File: src/modules/minio/minio.service.ts [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+  private async createBucketIfNotExists(bucketName: BucketName) {
+    const bucketExists = await this.minioClient.bucketExists(bucketName);
+
+    if (!bucketExists) {
+      await this.minioClient.makeBucket(bucketName);
+    }
+  }
+
+  async uploadFile(file: Express.Multer.File, bucketName = BucketName.Images) {
+    const fileName = `${Date.now()}-${file.originalname}`;
```
Comment (farrooh) [inline src/modules/minio/minio.service.ts:84]: (Optional) Using Date.now() is not really a bad idea, but it is better to use some other unique values like uuid or nanoid

**Review by farrooh — CHANGES_REQUESTED** [review]

File: src/modules/minio/minio.service.ts [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+      prevPolicy['Statement'].push(bucketPolicyRead.Statement[0]);
+    }
+
+    try {
+      await this.minioClient.setBucketPolicy(
+        bucketName,
+        JSON.stringify(prevPolicy),
+      );
+    } catch (err) {
+      console.error(err.message);
```
Comment (Ismoil-JS) [inline src/modules/minio/minio.service.ts:70]: remove console.log after finishing task

**Review by Ismoil-JS — CHANGES_REQUESTED** [review]
remove all console

File: src/modules/minio/minio.service.ts (hunk 5 above)
Comment (tiescl) [reply src/modules/minio/minio.service.ts]: I dont have console.logs

File: src/modules/minio/minio.service.ts (hunk 4 above)
Comment (tiescl) [reply src/modules/minio/minio.service.ts]: fixed

File: src/modules/minio/minio.service.ts (hunk 3 above)
Comment (tiescl) [reply src/modules/minio/minio.service.ts]: fixed

File: src/modules/minio/minio.service.ts (hunk 2 above)
Comment (tiescl) [reply src/modules/minio/minio.service.ts]: fixed

**Review by farrooh — APPROVED** [review]

**Review by Ismoil-JS — APPROVED** [review]

**Pull request merged**